### PR TITLE
Support pagination for resources loading

### PR DESCRIPTION
**Pull request by boajer**
https://oat-sa.atlassian.net/browse/CON-313

### CODE REVIEW

File: models/classes/media/LocalItemSource.php [hunk 1]
```diff
@@ -61,15 +61,16 @@ public function getItem()
...
      * @throws common_exception_Error
      */
-    public function getDirectory($parentLink = '', $acceptableMime = [], $depth = 1)
+    public function getDirectory(
+        string $parentLink = '',
+        array $acceptableMime = [],
+        int $depth = 1,
+        int $limit = 0,
+        int $offset = 0
+    ): array
```
Comment (boajer) [inline models/classes/media/LocalItemSource.php:73]: Proper pagination support should be introduced after migration to `flysystem` 2.0 @see https://flysystem.thephpleague.com/v2/docs/usage/directory-listings/ in order to have reliable filtering

File: actions/class.ItemContent.php [hunk 2]
```diff
@@ -36,29 +36,33 @@ class taoItems_actions_ItemContent extends tao_actions_CommonModule
...
-        $item = $this->getResource($itemUri);
+        $itemUri = $params['uri'];
+        $item = $this->getResource($params['uri']);
 
-        if (!$this->hasRequestParameter('lang')) {
+        if (!isset($params['lang'])) {
             throw new common_exception_MissingParameter('lang', __METHOD__);
         }
-        $itemLang = $this->getRequestParameter('lang');
+        $itemLang = $params['lang'];
```
Comment (shpran) [inline actions/class.ItemContent.php:60]: In your case you will call `getResource` but then if `lang` param is not exist - you will throw an error.
To avoid this we can move second check to the top and after both checks we will call `getResource` method.
Wdyt?
```
        if (!isset($params['uri'])) {
            throw new common_exception_MissingParameter('uri', __METHOD__);
        }

        if (!isset($params['lang'])) {
            throw new common_exception_MissingParameter('lang', __METHOD__);
        }

        $itemUri = $params['uri'];
        $item = $this->getResource($params['uri']);
        $itemLan = $params['lang'];
```

File: actions/class.ItemContent.php [hunk 3]
```diff
@@ -73,13 +77,17 @@ public function files()
                 $filters = array_map('trim', explode(',', $filterParameter));
             }
         }
-        $depth = $this->hasRequestParameter('depth') ? $this->getRequestParameter('depth') : 1;
+        $depth = isset($params['depth']) ? $params['depth'] : 1;
+        $limit = self::DEFAULT_PAGINATION_LIMIT;
```
Comment (shpran) [inline actions/class.ItemContent.php:81]: Do we need this variable?
Maybe we can use constant directly?

```
        $data = $asset->getMediaSource()->getDirectory(
            $asset->getMediaIdentifier(),
            $filters,
            $depth,
            self::DEFAULT_PAGINATION_LIMIT,
            $offset
        );
```

File: actions/class.ItemContent.php [hunk 4]
```diff
@@ -73,13 +77,17 @@ public function files()
...
+        $offset = isset($params['offset']) ? $params['offset'] : self::DEFAULT_PAGINATION_OFFSET;
```
Comment (shpran) [inline actions/class.ItemContent.php:82]:
```suggestion
        $offset = $params['offset'] ?? self::DEFAULT_PAGINATION_OFFSET;
```

**Review by shpran — CHANGES_REQUESTED** [review]
Looks good 👍🏻 
Minor comments/suggestions.

File: actions/class.ItemContent.php [hunk 5]
```diff
@@ -36,29 +37,34 @@ class taoItems_actions_ItemContent extends tao_actions_CommonModule
 {
     use OntologyAwareTrait;
 
+    private const DEFAULT_PAGINATION_LIMIT = 10;
```
Comment (mike-ionut-mihai-sandu-tao) [inline actions/class.ItemContent.php:40]: can we put this inside a config or try to take it from the env ?

File: actions/class.ItemContent.php [hunk 6]
```diff
@@ -36,29 +37,34 @@ class taoItems_actions_ItemContent extends tao_actions_CommonModule
...
+        $params = $this->getPsrRequest()->getQueryParams();
+
+        if (!isset($params['uri'])) {
             throw new common_exception_MissingParameter('uri', __METHOD__);
         }
-        $itemUri = $this->getRequestParameter('uri');
-        $item = $this->getResource($itemUri);
 
-        if (!$this->hasRequestParameter('lang')) {
+        if (!isset($params['lang'])) {
```
Comment (mike-ionut-mihai-sandu-tao) [inline actions/class.ItemContent.php:56]: i would use !empty because they might contain '' as  a value

**Review by mike-ionut-mihai-sandu-tao — COMMENTED** [review]
lgtm, left a few comments

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
LocalItemSourceTest require some changes, no?

File: actions/class.ItemContent.php [hunk 7]
```diff
@@ -36,29 +37,34 @@ class taoItems_actions_ItemContent extends tao_actions_CommonModule
...
+
     /**
-     * Returns a json encoded array describign a directory
+     * Returns a json encoded array describing a directory
      *
      * @throws common_exception_MissingParameter
-     * @return string
      */
-    public function files()
+    public function files(): void
```
Comment (gabrielfs7) [inline actions/class.ItemContent.php:48]: Void, if returns an array?

File: actions/class.ItemContent.php [hunk 8]
```diff
@@ -73,13 +79,20 @@ public function files()
...
+        $childrenLimit = $this->getPaginationLimit();
+        $childrenOffset = $params['childrenOffset'] ?? self::DEFAULT_PAGINATION_OFFSET;
+
         $resolver = new ItemMediaResolver($item, $itemLang);
-        $asset = $resolver->resolve($this->getRequestParameter('path'));
-        $data = $asset->getMediaSource()->getDirectory($asset->getMediaIdentifier(), $filters, $depth);
+        $asset = $resolver->resolve($params['path']);
+
+        $data = $asset->getMediaSource()->getDirectories(
+            new QueryObject($asset->getMediaIdentifier(), $filters, $depth, $childrenLimit, (int)$childrenOffset)
```
Comment (gabrielfs7) [inline actions/class.ItemContent.php:90]: I mentioned in the other PR to rename this object `QueryObject` to `DirectorySearchQuery` and not put it on tao-media-manager extension. 

Now we are coupling media manager with this extension. Wdyt?

File: actions/class.ItemContent.php [hunk 9]
```diff
@@ -271,4 +284,9 @@ private function getContentDetector(): ContentDetector
         /** @noinspection PhpIncompatibleReturnTypeInspection */
         return $this->getServiceLocator()->get(ContentDetector::class);
     }
+
+    private function getPaginationLimit(): int
```
Comment (gabrielfs7) [inline actions/class.ItemContent.php:288]: Why a method to get a private constant?

File: models/classes/media/LocalItemSource.php [hunk 10]
```diff
@@ -314,4 +267,67 @@ protected function getItemDirectory()
...
+        } else {
+            $directory = $itemDirectory;
+        }
+
+        $iterator = $directory->getFlyIterator();
+        $total = 0;
+        foreach ($iterator as $content) {
+            if ($content instanceof Directory) {
+                $children[] = $this->searchDirectories($itemDirectory->getRelPath($content), $acceptableMime,
+                    $depth - 1, $childrenLimit);
```
Comment (gabrielfs7) [inline models/classes/media/LocalItemSource.php:318]: Why $total is not incremented here?

I think this very method is too complex and should be in another class.

File: models/classes/media/LocalItemSource.php [hunk 11]
```diff
@@ -314,4 +267,67 @@ protected function getItemDirectory()
...
+            } else {
+                $fileInfo = $this->getInfoFromFile($content);
+                if (empty($acceptableMime) || in_array($fileInfo['mime'], $acceptableMime)) {
+                    $children[] = $fileInfo;
+                    $total++;
+                }
+            }
```
Comment (gabrielfs7) [inline models/classes/media/LocalItemSource.php:325]:
```suggestion
            if ($content instanceof Directory) {
                $children[] = $this->searchDirectories($itemDirectory->getRelPath($content), $acceptableMime,
                    $depth - 1, $childrenLimit);
            
                $total++;
                        
                continue; 
            }
                
            $fileInfo = $this->getInfoFromFile($content);
            
            if (empty($acceptableMime) || in_array($fileInfo['mime'], $acceptableMime)) {
                $children[] = $fileInfo;
                
                $total++;
            }
```

Reduce nested blocks / complexity. You do not need else and you have to increment total always.

File: models/classes/media/LocalItemSource.php [hunk 12]
```diff
@@ -57,69 +58,21 @@ public function getItem()
         return $this->item;
     }
 
+    public function getDirectories(QueryObject $params): array
+    {
+        return $this->searchDirectories($params->getParentLink(), $params->getFilter(), $params->getDepth(),
+            $params->getChildrenLimit());
```
Comment (gabrielfs7) [inline models/classes/media/LocalItemSource.php:64]: To follow PSR-12. The exception (which I do not agree personally) is only about array and lambda parameters

```suggestion
        return $this->searchDirectories(
            $params->getParentLink(), 
            $params->getFilter(), 
            $params->getDepth(),
            $params->getChildrenLimit()
        );
```

**Review by gabrielfs7 — CHANGES_REQUESTED** [review]
Thanks @boajer !

- Some changes required
- missing tests

File: actions/class.ItemContent.php (hunk 8 above)
Comment (boajer) [reply actions/class.ItemContent.php:90]: Did I miss something, but class is already located at tao-core, so no extra coupling there. 
BTW good idea about naming.

File: actions/class.ItemContent.php (hunk 7 above)
Comment (boajer) [reply actions/class.ItemContent.php:48]: Yes, method indeed returns void;  I will change miss-leading method description.

File: actions/class.ItemContent.php (hunk 8 above)
Comment (gabrielfs7) [reply actions/class.ItemContent.php:90]: Yes, the method is in tao-core already. You are right :)

File: actions/class.ItemContent.php (hunk 6 above)
Comment (bartlomiejmarszal) [inline actions/class.ItemContent.php]: We could move those to separate method.

File: migrations/Version202012161236262141_taoItems.php [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+use Doctrine\DBAL\Schema\Schema;
+use oat\tao\scripts\tools\migrations\AbstractMigration;
+use oat\taoItems\model\media\AssetTreeBuilder;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version202012161236262141_taoItems extends AbstractMigration
+{
+
```
Comment (gabrielfs7) [inline migrations/Version202012161236262141_taoItems.php:16]:
```suggestion
```

File: migrations/Version202012161236262141_taoItems.php [hunk 14]
```diff
@@ -0,0 +1,28 @@
+<?php
+
```
Comment (gabrielfs7) [inline migrations/Version202012161236262141_taoItems.php:2]: Missing doc

File: migrations/Version202012161236262141_taoItems.php [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version202012161236262141_taoItems extends AbstractMigration
+{
+
+    public function up(Schema $schema): void
+    {
+        $this->getServiceManager()->register(AssetTreeBuilder::SERVICE_ID, new AssetTreeBuilder([
+            AssetTreeBuilder::OPTION_PAGINATION_LIMIT => 15,
+        ]));
```
Comment (gabrielfs7) [inline migrations/Version202012161236262141_taoItems.php:21]: Bit hard to read what is parameter for what

```suggestion
        $this->getServiceManager()->register(
            AssetTreeBuilder::SERVICE_ID, 
            new AssetTreeBuilder(
                [
                    AssetTreeBuilder::OPTION_PAGINATION_LIMIT => 15,
                ]
            )
        );
```

File: models/classes/media/AssetTreeBuilder.php [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+use tao_helpers_Uri;
+
+class AssetTreeBuilder extends ConfigurableService
+{
+    public const SERVICE_ID = 'taoItems/AssetTreeBuilder';
+
+    public const OPTION_PAGINATION_LIMIT = 'pagination_limit';
+
+    private const DEFAULT_PAGINATION_LIMIT = 15;
+    public const DEFAULT_PAGINATION_OFFSET = 0;
```
Comment (gabrielfs7) [inline models/classes/media/AssetTreeBuilder.php:36]: maybe first public, then private?

File: migrations/Version202012161236262141_taoItems.php (hunk 14 above)
Comment (mike-ionut-mihai-sandu-tao) [reply migrations/Version202012161236262141_taoItems.php:2]: but this being a migration i do see it as absolutely required

File: models/classes/media/LocalItemSource.php [hunk 17]
```diff
@@ -57,69 +58,25 @@ public function getItem()
...
     /**
      * Get a array representing the tree of directory
      *
      * @see \oat\tao\model\media\MediaBrowser::getDirectory
-     * @param string $parentLink
-     * @param array $acceptableMime
-     * @param int $depth
-     * @return array
-     * @throws \FileNotFoundException
-     * @throws \tao_models_classes_FileNotFoundException
```
Comment (bartlomiejmarszal) [inline models/classes/media/LocalItemSource.php:69]: Why did you remove this?

File: actions/class.ItemContent.php [hunk 18]
```diff
@@ -24,70 +24,50 @@
...
-            } else {
-                if (preg_match('/\/\*/', $filterParameter)) {
-                    $this->logWarning('Stars mime type are not yet supported, filter "' . $filterParameter . '" will fail');
-                }
-                $filters = array_map('trim', explode(',', $filterParameter));
-            }
-        }
-        $depth = $this->hasRequestParameter('depth') ? $this->getRequestParameter('depth') : 1;
+        $itemLang = $params['lang'];
+        $path = $params['path'];
```
Comment (mike-ionut-mihai-sandu-tao) [inline actions/class.ItemContent.php:56]: do we need to check this or it is optional ?

File: actions/class.ItemContent.php [hunk 19]
```diff
@@ -266,9 +246,38 @@ protected function getMediaSource($urlPrefix, $item, $itemLang)
         return $asset->getMediaSource();
     }
 
+    private function buildFilters(array $params): array
```
Comment (mike-ionut-mihai-sandu-tao) [inline actions/class.ItemContent.php:249]: should this be extracted?

File: models/classes/media/LocalItemSource.php [hunk 20]
```diff
@@ -314,4 +271,73 @@ protected function getItemDirectory()
...
+        $label = rtrim($parentLink, '/');
+        if (strrpos($parentLink, '/') !== false && substr($parentLink, -1) !== '/') {
+            $label = substr($parentLink, strrpos($parentLink, '/') + 1);
+            $parentLink = $parentLink . '/';
+        }
+
+        if (in_array($parentLink, ['', '/'])) {
+            $label = $this->getItem()->getLabel();
+            $parentLink = '/';
+        }
```
Comment (bartlomiejmarszal) [inline models/classes/media/LocalItemSource.php:294]: Can we extract this to separate method?

**Review by mike-ionut-mihai-sandu-tao — COMMENTED** [review]
lgtm , left  a few comments

File: models/classes/media/LocalItemSource.php [hunk 21]
```diff
@@ -314,4 +271,73 @@ protected function getItemDirectory()
         }
         return $this->itemService->getItemDirectory($this->item, $this->lang);
     }
+
+    private function searchDirectories(
```
Comment (gabrielfs7) [inline models/classes/media/LocalItemSource.php:275]: Missing throws annotation

**Review by bartlomiejmarszal — COMMENTED** [review]
Great! Am I still not see any new tests?

File: migrations/Version202012161236262141_taoItems.php [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace oat\taoItems\migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use oat\tao\scripts\tools\migrations\AbstractMigration;
+use oat\taoItems\model\media\AssetTreeBuilder;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
```
Comment (gabrielfs7) [inline migrations/Version202012161236262141_taoItems.php:13]: I suppose this is not autogenerated anymore, right?

File: actions/class.ItemContent.php (hunk 18 above)
Comment (boajer) [reply actions/class.ItemContent.php:56]: Good catch, It's required, but original code has no validation on it; so adding it

File: actions/class.ItemContent.php (hunk 19 above)
Comment (boajer) [reply actions/class.ItemContent.php:249]: I think that helped to improve readability

File: models/classes/media/LocalItemSource.php (hunk 20 above)
Comment (boajer) [reply models/classes/media/LocalItemSource.php:294]: Yes, but there are already to many changes not directly related to the purpose of PR, as for code that needs to be backported at the 1.5 years old version, so I would avoid any additional refactoring.

**Review by mike-ionut-mihai-sandu-tao — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: models/classes/media/AssetTreeBuilder.php [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+
+declare(strict_types=1);
+
+namespace oat\taoItems\model\media;
+
+use oat\oatbox\service\ConfigurableService;
+use oat\tao\model\media\mediaSource\DirectorySearchQuery;
+use tao_helpers_Uri;
+
+class AssetTreeBuilder extends ConfigurableService
```
Comment (gabrielfs7) [inline models/classes/media/AssetTreeBuilder.php:29]: You are working in the test, right?

**Review by gabrielfs7 — APPROVED** [review]

File: actions/class.ItemContent.php [hunk 24]
```diff
@@ -266,9 +250,38 @@ protected function getMediaSource($urlPrefix, $item, $itemLang)
...
+        return $filters;
+    }
+
     private function getContentDetector(): ContentDetector
     {
         /** @noinspection PhpIncompatibleReturnTypeInspection */
         return $this->getServiceLocator()->get(ContentDetector::class);
     }
+
+    /** @noinspection PhpIncompatibleReturnTypeInspection */
```
Comment (bartlomiejmarszal) [inline actions/class.ItemContent.php:281]:
```suggestion
```

**Review by bartlomiejmarszal — APPROVED** [review]
LGTM

- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
